### DISCUSSION TITLE
merging two tables with same primary key

### DISCUSSION THREAD

**Discussion by rhcgn** [ClickHouse/ClickHouse#34286] [Q&A]
I have one table storing the hits with primary key 'request_id'. I have another table storing user hashes also with primary key 'request_id' (both are filled by different csv-files). User hases are the link to user data like age or zip code. Now I need to merge the two tables to connect user data with requests via external dictionary, which works well. But I have problems with merging the request data with the user hashes. Both tables are increasing very fast. I tried to import them both into same table, but aggeregation process took too long. I tried an external dictionary for the hash table with LAYOUT(DIRECT()) but then dictGet function eventually crashes with memory overflow. A real dictionary for hash table does not make sense to me as every data point is just needed once for exact one row in request table. Any ideas or best practices?

**Comment by den-crane** [comment]
>I tried an external dictionary for the hash table with LAYOUT(DIRECT()) 
>but then dictGet function eventually crashes with memory overflow.

It should not. It's a bug. Which CH version do you use?

**Reply by rhcgn** [reply]
I use version 21.12

**Reply by rhcgn** [reply]
But maybe it is because of Grouping:

```
create materialized view mv_customers to visits
as SELECT
       visit_id as visitID,
       upper(dictGet('dict_hashedmail','emailhash',request_id)) as hash,
       max(toInt64(dictGet('dict_custhash','age',hash))) AS age,
       min(dictGet('dict_custhash','age_group',hash)) AS age_group
FROM requests
WHERE dictHas('dict_hashedmail',request_id) AND hash != 'NULL'
GROUP BY visitID,hash
```

**Reply by rhcgn** [reply]
Also I tried LAYOUT(HASHED()) now for dictionary with user hashes. That worked fine. Even though I don't think it is a good idea to put all request ids in memory. Sooner or later this will also crash I guess.
